### PR TITLE
Serializes nil and symbol objects

### LINKED ISSUE #1: Preserve symbols instead of converting them to strings
The expected and actual values for an array of symbols is serialized to an array of string values because of

https://github.com/firstdraft/rspec-enriched_json/blob/e2e0e4b38eac30326ec96dd546f52ad41439ff4a/lib/rspec/enriched_json/expectation_helper_wrapper.rb#L37

```json
"details": {
  "expected": [
    "city",
    "state",
    "zip"
  ],
  "actual": "",
  "matcher_name": "RSpec::Matchers::DSL::Matcher",
  "diffable": false
}
```

I think this should be fixed.

The `exception>message` value retails them as an array of symbols.

```json
"message": "Expected output to match: '[:city, :state, :zip]', but was: ''."
``` 

This would be confusing to users.

### PR DESCRIPTION
**Pull request by swaathi**
Resolves #1 and #2

### Problem

There is no JSON equivalent of Ruby's Symbols and the NilClass. So when these values are present in the `expected` or `actual` key of the `details` object, they render out to their String format.

This causes the `expected` or `actual` values to look different than what's displayed in the `exception>message`.

### Solution

Serialize Symbol and NilClass objects. This provides the receiver with enough information to parse their JSON representation to Ruby.

### Sample test

```ruby
RSpec.describe "Simple match test" do
  it "compares with nil" do
    expect(nil).to eq(true)
  end

  it "compares with string nil" do
    expect("nil").to eq(true)
  end

  it "compares with symbol" do
    expect(:nil).to eq(true)
  end

  it "compares with string symbol" do
    expect(":nil").to eq(true)
  end

  it "compares array of symbols" do
    expect([:abc]).to eq([:def])
  end

  it "compares array of string symbols" do
    expect([":abc"]).to eq([":def"])
  end
end
```

Details object for the above test
```json
[
  {
    "expected": true,
    "actual": ":nil",
    "matcher_name": "RSpec::Matchers::BuiltIn::Eq",
    "diffable": true,
    "diff": "\n@@ -1 +1 @@\n-true\n+\":nil\"\n"
  },
  {
    "expected": true,
    "actual": {
      "class": "Symbol",
      "inspect": ":nil",
      "to_s": "nil"
    },
    "matcher_name": "RSpec::Matchers::BuiltIn::Eq",
    "diffable": true,
    "diff": "\n@@ -1 +1 @@\n-true\n+:nil\n"
  },
  {
    "expected": [
      {
        "class": "Symbol",
        "inspect": ":def",
        "to_s": "def"
      }
    ],
    "actual": [
      {
        "class": "Symbol",
        "inspect": ":abc",
        "to_s": "abc"
      }
    ],
    "matcher_name": "RSpec::Matchers::BuiltIn::Eq",
    "diffable": true,
    "diff": "\n@@ -1 +1 @@\n-[:def]\n+[:abc]\n"
  },
  {
    "expected": [":def"],
    "actual": [":abc"],
    "matcher_name": "RSpec::Matchers::BuiltIn::Eq",
    "diffable": true
  },
  {
    "expected": true,
    "actual": "nil",
    "matcher_name": "RSpec::Matchers::BuiltIn::Eq",
    "diffable": true,
    "diff": "\n@@ -1 +1 @@\n-true\n+\"nil\"\n"
  },
  {
    "expected": true,
    "actual": {
      "class": "NilClass",
      "inspect": "nil",
      "to_s": ""
    },
    "matcher_name": "RSpec::Matchers::BuiltIn::Eq",
    "diffable": true
  }
]
```

----

> [!IMPORTANT]
> Serialize `Symbol` and `NilClass` in JSON to maintain consistency in RSpec test outputs.
> 
>   - **Behavior**:
>     - Serialize `Symbol` and `NilClass` objects in `serialize_value()` in `expectation_helper_wrapper.rb` using `serialize_object()`.
>     - Ensures JSON representation can be parsed back to Ruby, maintaining consistency in `expected` and `actual` values.
>   - **Tests**:
>     - Add tests in `edge_cases_spec.rb` for serializing `nil` and `symbols`.
>     - Ensure graceful handling of circular references, matchers without `expected/actual`, and encoding issues.
>   - **Misc**:
>     - Add test cases in `demo.rb` for `eq` with array of symbols and `nil`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=firstdraft%2Frspec-enriched_json&utm_source=github&utm_medium=referral)<sup> for 3e2418ca1d170f30a6fa531d9b67d1fd6434ab60. You can [customize](https://app.ellipsis.dev/firstdraft/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: lib/rspec/enriched_json/expectation_helper_wrapper.rb [hunk 1]
```diff
@@ -29,14 +29,16 @@ def serialize_value(value, depth = 0)
...
+          when Numeric, TrueClass, FalseClass
             value
           when String
             unescape_string_double_quotes(
               truncate_string(value)
             )
           when Symbol
-            value.to_s
+            serialize_object(value)
+          when nil
```
Comment (jelaniwoods) [inline lib/rspec/enriched_json/expectation_helper_wrapper.rb:40]: To be consistent with the other cases what do you think about using `NilClass` instead?

```suggestion
          when NilClass
```

**Review by bpurinton — APPROVED** [review]
LGTM, I think this is an acceptable solution.

**Comment by bpurinton** [comment]
@swaathi, oops, I just checked the specs here too and saw:

```
Finished in 5.36 seconds (files took 0.12569 seconds to load)
31 examples, 3 failures

Failed examples:

rspec ./spec/edge_cases_spec.rb:18 # Edge case handling handles matchers without expected/actual methods
rspec ./spec/edge_cases_spec.rb:70 # Edge case handling handling Ruby literals serializes symbols
rspec ./spec/edge_cases_spec.rb:61 # Edge case handling handling Ruby literals serializes nil
```

**Review by jelaniwoods — CHANGES_REQUESTED** [review]
These tests fail in a codespace

image

**Comment by bpurinton** [comment]
> These tests fail in a codespace

Also locally

File: lib/rspec/enriched_json/expectation_helper_wrapper.rb (hunk 1 above)
Comment (swaathi) [reply lib/rspec/enriched_json/expectation_helper_wrapper.rb:40]: Yup, thought about this. But I suppose there was a reason it was `nil` before this change, so I left it as such.

I can spend some time and see if this change would affect anything.

**Comment by swaathi** [comment]
So these specs do not pass because they use the "live" version of the gem. Specs pass [here](https://github.com/firstdraft/rspec-enriched_json/blob/main/spec/diff_info_spec.rb) because they are written to a temp file and then run with a local build of the gem.

I think this should be left as is. It will pass when the gem is deployed.

**Comment by swaathi** [comment]
Updated the spec to run with the local build of the gem,

```ruby
def run_formatter_with_content(test_content)
  test_file = Tempfile.new(["test", ".rb"])
  test_file.write(test_content)
  test_file.flush

  output_file = Tempfile.new(["output", ".json"])

  # Run RSpec with our formatter
  system(
    "bundle", "exec", "rspec", test_file.path,
    "--format", "RSpec::EnrichedJson::Formatters::EnrichedJsonFormatter",
    "--out", output_file.path,
    "-r", "./lib/rspec/enriched_json",
    err: File::NULL,
    out: File::NULL
  )

  JSON.parse(File.read(output_file.path))
ensure
  test_file&.close
  test_file&.unlink
  output_file&.close
  output_file&.unlink
end

it "correctly serializes nil values" do
  test_content = <<~RUBY
    RSpec.describe "String diff" do
      it "compares strings" do
        expect("Alice").to eq(nil)
      end
    end
  RUBY

  output = run_formatter_with_content(test_content)
  details = output["examples"].first["details"]
  expect(details["expected"]["class"]).to eq("NilClass")
  expect(details["expected"]["inspect"]).to eq("nil")
  expect(details["expected"]["to_s"]).to eq("")
end
```

And it passes,

```Run options: include {locations: {"./spec/edge_cases_spec.rb" => [88]}}

Randomized with seed 12317

Edge case handling
  handles Ruby literals
    correctly serializes nil values

Finished in 0.36074 seconds (files took 0.27212 seconds to load)
1 example, 0 failures

Randomized with seed 12317
```

**Comment by bpurinton** [comment]
Nice!

**Comment by raghubetina** [comment]
This is really descriptive. I probably would have tried to use ActiveSupport for de/serialization (did you try that?), but your approach seems better suited for our needs. Nice!

**Comment by swaathi** [comment]
> I probably would have tried to use ActiveSupport for de/serialization (did you try that?), but your approach seems better suited for our needs.

No I wanted to keep things standard. We use `serialize_object` to handle Regex and other values.
